sys/include/net/sock/dtls.h

```diff
@@ -549,23 +557,27 @@ int sock_dtls_create(sock_dtls_t *sock, sock_udp_t *udp_sock,
 /**
  * @brief Creates a new DTLS session
  *
- * Initializes handshake process with a DTLS server  at @p ep.
+ * Start handshake process with a DTLS server  at @p ep.
  *
  * @param[in]  sock     DLTS sock to use
  * @param[in]  ep       Remote endpoint of the session
  * @param[out] remote   The created session, cannot be NULL
+ * @param[in]  timeout  Timeout to wait for handshake to finish.
+ *                      Returns immediately if 0.
+ *                      May be SOCK_NO_TIMEOUT to wait indefinitely until
+ *                      handshake complete.
  *
  * @return  0 on success
- * @return  -EAGAIN, if DTLS_HANDSHAKE_TIMEOUT is `0` and no data is available.
+ * @return  -EAGAIN, if @p timeout is `0` and handshake started.
```

Comment:
What is the use case for this again considering `sock_async`? As we only have `CONN_RDY` (a locally initialized handshake was finished) and `CONN_RECV` (a session was initialized by a remote peer) as events to deal with, I don't see much value for `timeout=0`. Starting a process that is not finished in the function it self and does not have a dedicated function to finish it, seems very inconsistent to me (calling `sock_dtls_recv()` should, from a user perspective, only be used to receive data IMHO, not to finish a handshake that was started as a side-effect of a certain parameter being set to a certain value).

Reply:
In gcoap, session create is usually called from the main thread by the user, no problem here. But sometimes, session create will be called from the listening thread (CON, peer restarts), initiating a new handshake, so the thread will be blocked until the handshake is finished or timeout. The problem here is we have a nested loop (session create loop in event loop) which will stop handling all other messages until the handshake is finished or timeout.

With `timeout=0` we can start the handshake and let the event loop handles receiving the handshake message. When the handshake message is complete, we can then resend the message that are due. I'm not sure how it is now with async merged but I assume it will be the same because we still have the event handler loop to handle receives.

Reply:
Then maybe we should have something like `accept()` in TCP, for accepting / finishing handshake requests?

Reply:
Hmmm then we'll need to manually inspect each packet for the DTLS record type, then emit an event that will require the application to call `accept()` if its a handshake packet. That seems like it'll work. Should this be included in this PR?

Reply:
Nvm, that's for later with async. For current `sock_dtls` how does `accept()` differ from recv? Should we reject all non-handshake packets?

Reply:
Again: the workflow for `async` and non-`async` should not differ in their core. If I call the functions in an order because I know I have to do it this way, or if they are called, because the events come in in that order, it shouldn't make a difference.

> Should we reject all non-handshake packets?

No, ideally they should be properly de-multiplexed, queued if need be. But I know, this is not easily possible with `tinydtls`. Let me think on this a bit.

Reply:
(the workflow should also not differ if the timeout is set or not)

Reply:
How about this: remove the timeout here and use `session_create` _always_ to just initialize the handshake. It than _has_ to be finished with `sock_dtls_recv()` for which we re-specify the scope as such.

Reply:
This also has the potential to make this API change less shaky: rename the new functionality `session_init()` and convert the `session_create()` to a wrapper for the proposed workflow.

Reply:
That sounds good. I'll go ahead and implement this to see how that works.

Reply:
Implemented this in 023cb680f. Simple test with the example application seems fine.